### PR TITLE
Move removal of duplicate indexes from fixture to Doctrine event listener

### PR DESCRIPTION
**Pull request by mollux**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [x]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 

#### Description:

When testing and working on the PHP 8 support (see https://github.com/mautic/mautic/pull/9969#issuecomment-1020322075), I noticed there are some fixtures that contain [DDL queries](https://www.geeksforgeeks.org/sql-ddl-dql-dml-dcl-tcl-commands/).

As Doctrine processes fixtures as 1 transaction that can be rolled back, and DDL queries force a commit when executed,
this results in an error when Doctrine tries to commit the transaction again.

in PHP < 8, this error was silently ignored, but in PHP >= 8 this error results in a ` There is no active transaction` PDOException.

One of the fixtures removes duplicate indexes on the `email_assets_xref` and `email_list_xref` tables.

This was only a partial (and with PHP 8 a breaking) fix, as there are multiple tables that have these duplicate indexes.  

Single column indexes that are the leftmost column of another index are obsolete.
That leftmost column is available to look up rows, so there is no need for an extra index.

See https://dev.mysql.com/doc/refman/5.7/en/multiple-column-indexes.html

That extra index is automatically created by Doctrine (e.g. for many to many join tables), and may be fixed in the future.
Luckily there are events we can subscribe to, so the duplicate indexes can be removed.

This PR adds the cleanup functionality for all duplicate indexes.
As side effect, all obsolete for many to many tables are deleted too

TO be able to test the impact, I used [`Percona tools duplicate key checker`](https://www.percona.com/doc/percona-toolkit/LATEST/pt-duplicate-key-checker.html)

https://github.com/percona/percona-toolkit/blob/3.x/bin/pt-duplicate-key-checker

**BEFORE**

```
# ########################################################################
# Summary of indexes
# ########################################################################

# Size Duplicate Indexes   8810
# Total Duplicate Indexes  40
# Total Indexes            618
```

**AFTER**

```
# ########################################################################
# Summary of indexes
# ########################################################################

# Size Duplicate Indexes   9497
# Total Duplicate Indexes  24
# Total Indexes            602
```

#### Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. install Mautic, and check if the many to many tables (you can recognise them by the `_xref` suffix) don't have an additional index for the first column in the primary key

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/11058"></a>

### CODE REVIEW

**Comment by mollux** [comment]
hm, not sure why the codecov fails, maybe because of the deleted test?

File: app/bundles/CoreBundle/EventListener/DoctrineEventsSubscriber.php [hunk 1]
```diff
@@ -125,6 +125,30 @@ public function postGenerateSchema(GenerateSchemaEventArgs $args)
...
+            }
         }
     }
+
+    /**
+     * Trim quotes from the identifier.
+     *
+     * @return string
+     */
+    private function trimQuotes(string $identifier)
```
Comment (escopecz) [inline app/bundles/CoreBundle/EventListener/DoctrineEventsSubscriber.php]:
```suggestion
     */
    private function trimQuotes(string $identifier): string
```

**Review by escopecz — COMMENTED** [review]
Looks great! One small suggestion.

Yes, the test coverage will decrease because a covered class is removed. I can't think of a way how to write a functional test for the postGenerateSchema method.

**Review by escopecz — APPROVED** [review]
When I execute `bin/console doctrine:schema:update --dump-sql` on the 4.x branch and compare it with the output on this PR then this is new:
```sql
     DROP INDEX IDX_BC68EA4619EB6921 ON mautic_oauth2_user_client_xref;
     DROP INDEX IDX_5379378BF639F774 ON mautic_campaign_leadlist_xref;
     DROP INDEX IDX_F0013AE3F639F774 ON mautic_campaign_form_xref;
     DROP INDEX IDX_14FE129BF639F774 ON mautic_campaign_leads;
     DROP INDEX IDX_11DC9DF2A832C1C9 ON mautic_email_list_xref;
     DROP INDEX IDX_39CFAB07A832C1C9 ON mautic_email_assets_xref;
     DROP INDEX IDX_B2794B6EB9FC8874 ON mautic_lead_lists_leads;
     DROP INDEX IDX_47BF800655458D ON mautic_lead_ips_xref;
     DROP INDEX IDX_BF8E2D1055458D ON mautic_lead_tags_xref;
     DROP INDEX IDX_CBE16758979B1AD6 ON mautic_companies_leads;
     DROP INDEX IDX_8FCED148EF1A9D84 ON mautic_push_notification_list_xref;
     DROP INDEX IDX_187896B8B42D874D ON mautic_channel_url_trackables;
     DROP INDEX IDX_7F02E38E71F7E88B ON mautic_point_lead_event_log;
     DROP INDEX IDX_4CF01FBDC028CEA2 ON mautic_point_lead_action_log;
     DROP INDEX IDX_913BA9C5BD5C7E60 ON mautic_sms_message_list_xref;
     DROP INDEX IDX_840FFA552298D193 ON mautic_stage_lead_action_log;
     DROP INDEX IDX_2AD4D584CE1C902 ON mautic_monitoring_leads;
```
It looks good! 👍

**Comment by RCheesley** [comment]
@mollux looks like CSFixer is failing as it's looking for `app/bundles/InstallBundle/InstallFixtures/ORM/RemoveDuplicateIndexData.php` and not finding it - wondering if this is expected?

I think we are OK with this PR if that's a false positive, given that @escopecz has reviewed the code and tested the outcome.

**Comment by escopecz** [comment]
I think this is a flaw in our CI that it asks CS fixer to fix only files that have been changed. For some reason the removed file is in the list too:
```
for file in app/bundles/CoreBundle/EventListener/DoctrineEventsSubscriber.php app/bundles/InstallBundle/Config/config.php app/bundles/InstallBundle/InstallFixtures/ORM/RemoveDuplicateIndexData.php app/bundles/InstallBundle/Tests/InstallFixtures/ORM/RemoveDuplicateIndexDataTest.php; do
      if [[ $file == *.php ]]; then
        cs_fix_files="${cs_fix_files} $file"
      fi
    done
    if [[ $cs_fix_files ]]; then
      bin/php-cs-fixer fix --config=.php-cs-fixer.php -v --dry-run --using-cache=no --show-progress=dots --diff $cs_fix_files
    fi
```
I'm merging this. It shouldn't cause troubles when merged.

**Pull request merged**